### PR TITLE
RES-2372: Add profiler and GPU Optimizations to dendrite experiments

### PR DESCRIPTION
**Pull request by lscheinkman**
This PR adds a mixin enabling the newest torch.profiler (requires torch>=1.8.1) and a few optimizations.
Depends on https://github.com/numenta/nupic.torch/pull/81

### CODE REVIEW

File: nupic/research/frameworks/vernon/mixins/torch_profiler.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+from nupic.research.frameworks.vernon.experiments import SupervisedExperiment
+
+__all__ = ["TorchProfilerMixin", "inject_torch_profiler_mixin"]
+
+
+class TorchProfilerMixin:
+    """
+    Mixin class enabling profiling via pytorch's native profiler.
+    See ahttps://pytorch.org/docs/stable/profiler.html
```
Comment (karangrewal) [inline nupic/research/frameworks/vernon/mixins/torch_profiler.py:35]: Typo.

**Review by karangrewal — APPROVED** [review]

File: nupic/research/frameworks/dendrites/functional/apply_dendrites.py [hunk 2]
```diff
@@ -65,7 +67,24 @@ def dendritic_bias_1d(y, dendrite_activations):
     return dendrite_output(y + winning_activations, indices)
 
 
-def dendritic_gate_1d(y, dendrite_activations, indices=None):
+@torch.jit.script
+def gather_activations(dendrite_activations, indices):
+    """
+    Gathers dendritic activations from the given indices.
+
```
[the remark this answers is not in the document]
Comment (lscheinkman) [reply nupic/research/frameworks/dendrites/functional/apply_dendrites.py:74]: torch.jit.script compiles the file in order and `dendritic_gate_1d` is calling it, so it needs to be defined before `dendritic_gate_1d`

File: nupic/research/frameworks/dendrites/model_utils.py [hunk 3]
```diff
@@ -106,12 +111,15 @@ def train_dendrite_model(
...
-        data = data.to(device)
-        target = target.to(device)
         if context is not None:
-            context = context.to(device)
+            context = context.to(device, non_blocking=async_gpu)
+
+        # FIXME: Pytorch 1.7: Replace with optimizer.zero_grad(set_to_none=True)
+        # optimizer.zero_grad(set_to_none=True)
+        for group in optimizer.param_groups:
+            for p in group["params"]:
```
[the remark this answers is not in the document]
Comment (lscheinkman) [reply nupic/research/frameworks/dendrites/model_utils.py:120]: `optimizer.zero_grad(set_to_none=True) ` is not available in torch 1.6.

[the remark this answers is not in the document]
Comment (lscheinkman) [reply nupic/research/frameworks/dendrites/model_utils.py:120]: The original `optimizer.zero_grad` will call `p.grad.zero_()`  using more memory and also performing the optimizer step on zero gradients. The optimizer usually skips the step if the gradient is `None` but it will not skip if the gradient is zero
see
- https://pytorch.org/tutorials/recipes/recipes/tuning_guide.html#use-parameter-grad-none-instead-of-model-zero-grad-or-optimizer-zero-grad 
- https://pytorch.org/docs/master/generated/torch.optim.Optimizer.zero_grad.html#torch.optim.Optimizer.zero_grad
